dev/MenuBar/MenuBarItem.cpp

```diff
@@ -260,13 +260,41 @@ void MenuBarItem::OpenFlyoutFrom(FlyoutLocation location)
         CloseMenuFlyout();
         if (location == FlyoutLocation::Left)
         {
-            winrt::get_self<MenuBarItem>(menuBar.Items().GetAt(((index - 1) + menuBar.Items().Size()) % menuBar.Items().Size()))->ShowMenuFlyout();
+            if (const auto item = GetNextItem(index, -1).try_as<MenuBarItem>())
+            {
+                item->ShowMenuFlyout();
+            }
         }
         else
         {
-            winrt::get_self<MenuBarItem>(menuBar.Items().GetAt((index + 1) % menuBar.Items().Size()))->ShowMenuFlyout();
+            if (const auto item = GetNextItem(index, +1).try_as<MenuBarItem>())
+            {
+                item->ShowMenuFlyout();
+            }
+        }
+    }
+}
+
+winrt::MenuBarItem MenuBarItem::GetNextItem(int index, int direction)
+{
+    if (const auto menuBar = m_menuBar.get())
+    {
+        const int itemsCount = menuBar.Items().Size();
+        // index + direction might be negative so account for that by adding itemsCount.
+        int itemIndex = (index + direction + itemsCount) % itemsCount;
+        while (!menuBar.Items().GetAt(itemIndex).IsEnabled())
```

Comment:
>IsEnabled [](start = 49, length = 9)

Do we also need to check for IsTabStop?

Reply:
I think so, yes. I've updated the check to include tabstop as criteria.

Reply:
Now that I'm reading this again, why do we need the item to be a tabstop?

---
In reply to: [536671264](https://github.com/microsoft/microsoft-ui-xaml/pull/3555#discussion_r536671264) [](ancestors = 536671264)

Reply:
Developers might have set IsTabStop to false manually, focusing it through keyboard navigation would then be a contradiction to the request of the MenuItem not being a tab stop.

Reply:
There are other things that can effect if an item is focusable.  I think the only real way to determine if an item is focusable is to try and set focus on it :/

@austin-lamb is that right?

Reply:
Unfortunately it's true that we don't have an IsFocusable property publicly, but I could tell you the things you'd want to look for if that's ok.  It means that in theory if we change the rules in the future this could get out of sync, but we haven't changed the rules in several years (the last time was when we introduced AllowFocusWhenDisabled).

@StephenLPeters - do you want me to look up the internal rules on focusability and it can be coded in here too?

Note that the "try to set focus" option isn't great because it means that you'll cause a bunch of focus events like [Getting|Losing|Got|Lost]Focus to fire, potentially when you don't want to.  This also affects what gets read out to an accessibility tool like Narrator.  Unless you'd just try to set focus and stop the first time focus successfully moves?  In that case, then "just trying to set focus" seems good and simple.

Reply:
The idea is to find the first enabled and focusable control in a given direction (left or right). @Austin-Lamb Do I understand this correctly that try move focus is fine here then? Could trying to move the focus multiple times cause any issues?

Reply:
If you try to move focus and it fails to move then that should be mostly benign.  A user might be signed up for the [Getting|Losing]Focus events and decide to cancel (via the eventArgs.TrySetCanceled(), for example) which could cause some event spam, but that is probably pretty uncommon.

In looking at your code, you're trying to find the item to call ShowMenuFlyout() on it - does that move focus to the item as well?  If so, then I think just calling menuBar.Items().GetAt(itemIndex).Focus(FocusState::Programmatic) would be the right thing to do here.  This will respect IsEnabled, IsTabStop, AllowFocusWhenDisabled, Visibility (and visibility of all ancestors) and all the other requirements for focusability.  If Focus() returns false, then we could not focus that item so you can keep looking for the next candidate.

Reply:
Thank you @Austin-Lamb that works and is a lot simpler! Since there don't seem to be much of a downside, I think this is just the best and easiest way moving forward here.

Reply:
Hooray, glad it's working!  Consider renaming the function since "GetNextItem" doesn't suggest it also manipulates focus, IMO.  Perhaps "GetAndFocusNextItem"?

Reply:
Good point, the name doesn't entirely reflect what the function. Opted for "FocusAndReturnNextFocusableItem" to indicate the order of steps and what exactly will be returned.